lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs

```diff
@@ -891,139 +891,134 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
-                in
-                W.Coin $ (round perByte) * (extraBytes + scriptIntegrityHashBytes)
-
-            fromCardanoLovelace (Cardano.Lovelace l) = Coin.unsafeFromIntegral l
-
-            selectionConstraints = SelectionConstraints
-                { assessTokenBundleSize =
-                    view #assessTokenBundleSize $
-                    tokenBundleSizeAssessor txLayer $
-                    pp ^. (#txParameters . #getTokenBundleMaxSize)
-                , certificateDepositAmount =
-                    view #stakeKeyDeposit pp
-                , computeMinimumAdaQuantity =
-                    view #txOutputMinimumAdaQuantity
-                        (constraints txLayer era pp)
-                , isBelowMinimumAdaQuantity =
-                    view #txOutputBelowMinimumAdaQuantity
-                        (constraints txLayer era pp)
-                , computeMinimumCost = \skeleton -> mconcat
-                    [ feePadding
-                    , fromCardanoLovelace fee0
-                    , calcMinimumCost txLayer era pp
-                        (defaultTransactionCtx
-                            { txPaymentCredentialScriptTemplate = mScriptTemplate
-                            , txPlutusScriptExecutionCost =
-                                txPlutusScriptExecutionCost })
-                        skeleton
-                    ] `Coin.difference` boringFee
-                , computeSelectionLimit = \_ -> NoLimit
-                , maximumCollateralInputCount = unsafeIntCast @Natural @Int $
-                    case recentEra @era of
-                        RecentEraBabbage ->
-                            getField @"_maxCollateralInputs" ledgerPP
-                        RecentEraConway ->
-                            getField @"_maxCollateralInputs" ledgerPP
-
-                , minimumCollateralPercentage = case recentEra @era of
-                    -- case-statement avoids "Overlapping instances" problem.
-                    -- May be avoidable with ADP-2353.
+        flip evalRand (stdGenFromSeed seed)
+            $ runExceptT
+            $ performSelection selectionConstraints selectionParams
+      where
+        txPlutusScriptExecutionCost = maxScriptExecutionCost pp redeemers
+        colReq =
+            if txPlutusScriptExecutionCost > W.Coin 0
+                then SelectionCollateralRequired
+                else SelectionCollateralNotRequired
+
+        (positiveBundle, negativeBundle) = posAndNegFromCardanoValue balance
+        TokenBundle positiveAda positiveTokens = positiveBundle
+        TokenBundle negativeAda negativeTokens = negativeBundle
+        adaInOutputs = F.foldMap (TokenBundle.getCoin . view #tokens) outs
+        tokensInOutputs = F.foldMap (TokenBundle.tokens . view #tokens) outs
+        selectedBalance = UTxOSelection.selectedBalance utxoSelection
+        tokensInInputs = TokenBundle.tokens selectedBalance
+        adaInInputs = TokenBundle.getCoin selectedBalance
```

Comment:
We could do:
```suggestion
        TokenBundle adaInInputs tokensInInputs 
            = UTxOSelection.selectedBalance utxoSelection
```
and same for outputs, which I think is clearer by
- being consistent with line `{positive,negative}Bundle` above
- avoiding the temporary `selectedBalance` definition.